### DISCUSSION TITLE
Overload resolution should be smarter about unambiguous Methods.

### DISCUSSION THREAD

**Discussion by TonyValenti** [dotnet/csharplang#1550] [Language Ideas]
1.  I understand that generic constraints are encoded as metadata and because of that, you can't have two methods in the same class that vary only by constraints.
2.  I believe you should be able to have two extension methods in different classes that vary only by constraints and C# should be able to resolve the differences.  The following code doesn't work but I believe it should:
````
public static class Extensions1 {
  public static T DoSomething(this T This) where T:class {
  ....
  }
}

public static class Extensions2 {
  public static T DoSomething(this T This) where T:Person {
  ....
  }
}

public static void Test(){
  var P = new Person()
   .DoSomething() //This causes the issue.
   ;
}

````

**Comment by HaloFour** [comment]
Should this not have been resolved by "bestest betterness" or did that not apply to extension methods?

**Comment by HaloFour** [comment]
Actually, "bestest betterness" only removes candidates that don't meet the constraints.  It doesn't attempt to determine a "best fit" set of constraints.

**Comment by Joe4evr** [comment]
> Actually, "bestest betterness" only removes candidates that don't meet the constraints. It doesn't attempt to determine a "best fit" set of constraints.

Correct. The new rules *do* allow overloading by generic constraint now, provided that the constraints don't overlap. Ergo, between `where T : class` and `where T : struct`. But since `where T : Person` implies `where T : class`, the constraints overlap and thus your call is **not** unambiguous.
